code/game/objects/structures/doors/_door.dm

```diff
@@ -0,0 +1,195 @@
+
+	var/changing_state = FALSE
+	var/icon_base
+	var/datum/lock/lock
+
+/obj/structure/door/Initialize()
+	. = ..()
+	if(!istype(material))
+		return INITIALIZE_HINT_QDEL
+	if(lock)
+		lock = new /datum/lock(src, lock)
+	if(!icon_base)
+		icon_base = material.door_icon_base
+	update_nearby_tiles(need_rebuild=TRUE)
+
+/obj/structure/door/Destroy()
+	update_nearby_tiles()
+	QDEL_NULL(lock)
+	return ..()
+
+/obj/structure/door/get_material_health_modifier()
+	. = 10
+
+/obj/structure/door/on_update_icon()
+	..()
+	if(density)
+		icon_state = "[icon_base]"
+	else
+		icon_state = "[icon_base]open"
+
+/obj/structure/door/proc/post_change_state()
+	update_nearby_tiles()
+	update_icon()
+	changing_state = FALSE
+
+/obj/structure/door/attack_hand(var/mob/user)
+	return density ? open() : close()
+
+/obj/structure/door/proc/close()
+	set waitfor = 0
+	if(!can_close())
+		return FALSE
+	flick("[icon_base]closing", src)
+	playsound(src.loc, material.dooropen_noise, 100, 1)
+
+	changing_state = TRUE
+	sleep(1 SECOND)
+	set_density(TRUE)
+	set_opacity(initial(opacity) && material.opacity > 0.5)
+	post_change_state()
+	return TRUE
+
+/obj/structure/door/proc/open()
+	set waitfor = 0
+	if(!can_open())
+		return FALSE
+	flick("[icon_base]opening", src)
+	playsound(src.loc, material.dooropen_noise, 100, 1)
+	
+	changing_state = TRUE
+	sleep(1 SECOND)
+	set_density(FALSE)
+	set_opacity(FALSE)
+	post_change_state()
+	return TRUE
+
+/obj/structure/door/proc/can_open()
+	if(lock && lock.isLocked())
+		return FALSE
+	return density && !changing_state
+
+/obj/structure/door/proc/can_close()
+	return !density && !changing_state
+
+/obj/structure/door/examine(mob/user, distance)
+	. = ..()
+	if(distance <= 1 && lock)
+		to_chat(user, SPAN_NOTICE("It appears to have a lock."))
+
+/obj/structure/door/attack_ai(mob/user)
+	if(Adjacent(user) && isrobot(user))
+		return attack_hand(user)
+
+/obj/structure/door/ex_act(severity)
+	take_damage(100 - (severity * 30))
+
+/obj/structure/door/can_repair(var/mob/user)
+	. = ..()
+	if(. && !density)
+		to_chat(user, SPAN_WARNING("\The [src] must be closed before it can be repaired."))
+		return FALSE
+
+/obj/structure/door/attackby(obj/item/I, mob/user)
+
+	add_fingerprint(user, 0, I)
+
+	if((user.a_intent == I_HURT && I.force) || istype(I, /obj/item/stack/material))
+		return ..()
+
+	if(lock)
+		if(istype(I, /obj/item/key))
+			if(!lock.toggle(I))
+				to_chat(user, SPAN_WARNING("\The [I] does not fit in the lock!"))
+			return TRUE
+		if(lock.pick_lock(I,user))
+			return TRUE
+		if(lock.isLocked())
+			to_chat(user, SPAN_WARNING("\The [src] is locked!"))
+		return TRUE
+
+	if(istype(I,/obj/item/material/lock_construct))
+		if(lock)
+			to_chat(user, SPAN_WARNING("\The [src] already has a lock."))
+		else
+			var/obj/item/material/lock_construct/L = I
+			lock = L.create_lock(src,user)
+		return
+
+	if(density)
+		open()
+	else
+		close()
+
+/obj/structure/door/CanPass(atom/movable/mover, turf/target, height=0, air_group=0)
```

Comment:
May need some flag set for this to work properly with ZAS; I have no clue about it, though.

Reply:
It seems to work correctly in testing, and this was copied from inflatables.